components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/sequence/impl/DefaultStepBasedSequenceHandler.java

```diff
@@ -427,6 +447,74 @@ protected void handlePostAuthentication(HttpServletRequest request,
         if (!authenticatedUserAttributes.isEmpty()) {
             sequenceConfig.getAuthenticatedUser().setUserAttributes(authenticatedUserAttributes);
         }
+        if (impersonatedUser != null) {
+            sequenceConfig.getAuthenticatedUser().setImpersonatedUser(impersonatedUser);
+            // Reset the user attributes returned from federate IdP if the requested claims are not empty.
+            if (!selectedRequestedClaims.isEmpty()) {
+                sequenceConfig.getAuthenticatedUser().getImpersonatedUser()
+                        .setUserAttributes(Collections.unmodifiableMap(new HashMap<>()));
+            }
+            if (isSPStandardClaimDialect(context.getRequestType()) && impersonatedUserAttributes.isEmpty()) {
+                sequenceConfig.getAuthenticatedUser().getImpersonatedUser()
+                        .setUserAttributes(impersonatedUserAttributes);
+            }
+            if (!impersonatedUserAttributes.isEmpty()) {
+                sequenceConfig.getAuthenticatedUser().getImpersonatedUser()
+                        .setUserAttributes(impersonatedUserAttributes);
+            }
+        }
+    }
+
+    private ImpersonatedUser getImpersonatedUser(HttpServletRequest request, AuthenticationContext context,
+                                                 AuthenticatedUser impersonatingActor)
+            throws FrameworkException {
+
+        ImpersonatedUser impersonatedUser = null;
+        // Get subject parameter from request.
+        String requestedSubject = extractFromQueryParams(request, context);
+        if (requestedSubject == null) {
+            // Get the subject from session context.
+            SessionContext sessionContext = FrameworkUtils.getSessionContextFromCache(context.getSessionIdentifier(),
+                    context.getLoginTenantDomain());
+            if (sessionContext != null && sessionContext.getImpersonatedUser() != null) {
```

Comment:
Lets move this check to the calling method

Reply:
addressed with 2af5ac045b6323d153825d3712f2801cd836ddd4